storage: reconcile manual splitting with automatic merging

jeffrey-xiao opened the pull request:
Follows the steps outlined in #37487 to reconcile manual splitting with automatic merging. This PR includes the changes needed at the storage layer. The sticky bit indicating that a range is manually split is added to the range descriptor.

jeffrey-xiao committed:
storage: remove reliance on disabling merge queue in tests

Remove all logic that disables merge queue from running in order to
preserve splits in tests since the splits should be preserved if they
were marked as manual splits.

The error outputted when attemping to split while the merge queue is
enabled is now gated behind VersionStickyBit.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37506)
<!-- Reviewable:end -->

nvb reviewed (commented):
Nice job getting this out so quickly! I left a large number of comments, but don't take that as a bad sign -- this is definitely on the right track.

> Reviewed 7 of 7 files at r1, 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-:-LetU4QJ2aUlITAH2DWf:b-8gfkdf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jeffrey-xiao)

___
*[pkg/internal/client/batch.go, line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Leqxns-3GwlEiMQ6N-y:-Leqxns-3GwlEiMQ6N-z:b-qb457x) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/internal/client/batch.go#L606)):*
> ```Go
> 		},
> 		SplitKey: splitKey,
> 		Manual:   true,
> ```

Leave a comment about why we're setting this.
___
*[pkg/kv/split_test.go, line 295 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetRhh66BO-GiD8IInH:-LetRhh66BO-GiD8IInI:boa6p5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/kv/split_test.go#L295)):*
> ```Go
> }
> 
> func TestRangeSplitsAlreadySplitStickyBit(t *testing.T) {
> ```

Give this a comment.
___
*[pkg/roachpb/api.proto, line 639 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetRlH45vxo8CJ_-ufw:-LetRlH45vxo8CJ_-ufx:boa6p5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/roachpb/api.proto#L639)):*
> ```protobuf
>   RequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
>   bytes split_key = 2 [(gogoproto.casttype) = "Key"];
>   bool manual = 3;
> ```

Give this a comment.
___
*[pkg/sql/split_test.go, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetS-wu4OxkASGe08u6:-LetS-wu4OxkASGe08u7:bu2xec3) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa0635be1ef7bfe03648a28c8e5c1c921ef5203c/pkg/sql/split_test.go#L61)):*
> ```Go
> 	r.Exec(t, "ALTER TABLE d.t SPLIT AT VALUES (1, 'a')")
> 
> 	// Prevent the merge queue from immediately discarding our splits.
> ```

This comment is echoed all over the place. I think we can remove anything that says this.
___
*[pkg/storage/client_merge_test.go, line 3318 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetSFL62zOZwacucX0c:-LetSFL62zOZwacucX0d:bf6qaye) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/client_merge_test.go#L3318)):*
> ```Go
> 	rhsEndKey := roachpb.RKey("c")
> 	for _, k := range []roachpb.RKey{lhsStartKey, rhsStartKey, rhsEndKey} {
> 		split(t, k.AsRawKey(), false)
> ```

Our style guide [discusses](https://github.com/cockroachdb/cockroach/blob/master/docs/style.md#inline-comments-for-arguments-to-function-calls) adding inline comments for constant arguments to functions. So this would become:
```
split(t, k.AsRawKey(), false /* manual */)
```

Let's make this change here and below.
___
*[pkg/storage/client_merge_test.go, line 3419 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetS_9f-KBUaWAb90o3:-LetS_9f-KBUaWAb90o4:b-5s3on1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/client_merge_test.go#L3419)):*
> ```Go
> 	})
> 
> 	t.Run("sticky bit", func(t *testing.T) {
> ```

Could you add a TODO right above this to add a subtest for considering load when making merging decisions? It would be nice to test this logic:
https://github.com/cockroachdb/cockroach/blob/de3ac9322ffc2fa0e35ccc94300e355ea3cdc6c5/pkg/storage/merge_queue.go#L258-L269
___
*[pkg/storage/client_merge_test.go, line 3424 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetT0Oe5g5O6jsy-PSK:-LetT0Oe5g5O6jsy-PSL:b-wtfxvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/client_merge_test.go#L3424)):*
> ```Go
> 		verifyUnmerged(t)
> 
> 		// Perform manual merge and verify that no merge occurred
> ```

Idiomatic Go uses sentences (with proper punctuation) for all comments: https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences.
___
*[pkg/storage/client_merge_test.go, line 3430 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetTAEcE4fEuPrNYg_8:-LetTAEcE4fEuPrNYg_9:bc7yi2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/client_merge_test.go#L3430)):*
> ```Go
> 		verifyUnmerged(t)
> 
> 		// Delete sticky bit and verify that merge occurs
> ```

Add a TODO to hook this up to the same mechanism that `ALTER TABLE ... UNSPLIT AT` uses when we eventually add that.
___
*[pkg/storage/merge_queue.go, line 213 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetTZptCxoq1iIiqI-t:-LetTZptCxoq1iIiqI-u:boa6p5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/merge_queue.go#L213)):*
> ```Go
> }
> 
> func (mq *mergeQueue) isManualSplit(ctx context.Context, splitKey roachpb.RKey) (bool, error) {
> ```

Give this a comment.
___
*[pkg/storage/merge_queue.go, line 313 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetTKOU4nqTsf-VtAUj:-LetTKOU4nqTsf-VtAUk:b-qx5jqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/merge_queue.go#L313)):*
> ```Go
> 	}
> 	if manualSplit {
> 		log.VEventf(ctx, 2, "skipping merge: ranges were manually split")
> ```

Please add a TODO to consider returning a "purgatory" error in this case to avoid repeatedly processing ranges that can't be merged.
___
*[pkg/storage/replica_command.go, line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetPFxX8rkJdoEf93du:-LetPFxX8rkJdoEf93dv:b-aqefmv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/storage/replica_command.go#L226)):*
> ```Go
> 		}
> 		log.Event(ctx, "range already split")
> 		// Even if the range is already split, we should still set the sticky bit
> ```

This is missing a `if args.Manual {` condition.

EDIT: it looks like you added this in the next commit. Let's rebase this fix back to the first one. Also, let's make sure that some test would have failed without this and without the corresponding condition in the main split transaction.
___
*[pkg/storage/replica_command.go, line 228 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetQQZQ5CBGcQxH5n1L:-LetQQZQ5CBGcQxH5n1M:b5scnu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/storage/replica_command.go#L228)):*
> ```Go
> 		// Even if the range is already split, we should still set the sticky bit
> 		err := r.store.DB().Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
> 			b := txn.NewBatch()
> ```

This transaction can currently race with a merge to leak sticky bits. In order to serialize with any merge transactions, I think we'll want to first read the range descriptor and make sure that it's equal to the provided `desc`. If not, we'll want to retry (see how `Replica.AdminSplit` handles `ConditionFailedError`s).

For reference, the split transaction below prevents these kinds of races by performing conditional puts on the range descriptor. We're not changing the range descriptor here, so there's no need to perform a CPut, but we'll want similar retry behavior.
___
*[pkg/storage/replica_command.go, line 319 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetRMP1DoX2MprXvgFd:-LetRMP1DoX2MprXvgFe:b84niv6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/storage/replica_command.go#L319)):*
> ```Go
> 		// Add sticky bit
> 		if args.Manual {
> 			b.Put(keys.SplitStickyBitKey(rightDesc.StartKey), true)
> ```

Where did we land on whether this should be a timestamp or a boolean?
___
*[pkg/storage/replica_command.go, line 549 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetTqwU866csTSv9sEe:-LetTqwU866csTSv9sEf:b-5c59sn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/storage/replica_command.go#L549)):*
> ```Go
> 
> func waitForApplication(
> 	ctx context.Context,
> ```

Let's get rid of this code movement. It changed in commit 1 and was later reverted. That could trip up a future `git blame`, so it's worth fixing in the first commit.


<!-- Sent from Reviewable.io -->

jeffrey-xiao committed:
storage: make merge queue respect sticky bit

When ranges are being processed in the merge queue, the sticky bit is
checked on the RHS of the merge operation. If the sticky bit is set,
then the merge does not go through.

Release note: None

jeffrey-xiao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37506)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jeffrey-xiao, and @nvanbenschoten)

___
*[pkg/storage/replica_command.go, line 319 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetRMP1DoX2MprXvgFd:-LevqzFP0lG38OXairsF:bvsoavl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/storage/replica_command.go#L319)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Where did we land on whether this should be a timestamp or a boolean?
</blockquote></details>

I think it depends if we want to support temporarily not merging manual splits. We would also need to consider how this time frame can be expressed on the SQL side.


<!-- Sent from Reviewable.io -->

jeffrey-xiao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37506)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jeffrey-xiao, and @nvanbenschoten)

___
*[pkg/internal/client/batch.go, line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Leqxns-3GwlEiMQ6N-y:-LewOJsiCEHDXStQlnXG:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/internal/client/batch.go#L606)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Leave a comment about why we're setting this.
</blockquote></details>

Done
___
*[pkg/kv/split_test.go, line 295 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetRhh66BO-GiD8IInH:-LewMmUU-2qQtPqTL7fc:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/kv/split_test.go#L295)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Give this a comment.
</blockquote></details>

Done
___
*[pkg/roachpb/api.proto, line 639 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetRlH45vxo8CJ_-ufw:-LewN7zYCbi9uHJMfCs0:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/roachpb/api.proto#L639)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Give this a comment.
</blockquote></details>

Done
___
*[pkg/storage/client_merge_test.go, line 3318 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetSFL62zOZwacucX0c:-LewN6Ai3MPFgH-QmhT9:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/client_merge_test.go#L3318)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Our style guide [discusses](https://github.com/cockroachdb/cockroach/blob/master/docs/style.md#inline-comments-for-arguments-to-function-calls) adding inline comments for constant arguments to functions. So this would become:
```
split(t, k.AsRawKey(), false /* manual */)
```

Let's make this change here and below.
</blockquote></details>

Done
___
*[pkg/storage/client_merge_test.go, line 3419 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetS_9f-KBUaWAb90o3:-LewN4mr0f9GdZ9lTz2m:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/client_merge_test.go#L3419)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you add a TODO right above this to add a subtest for considering load when making merging decisions? It would be nice to test this logic:
https://github.com/cockroachdb/cockroach/blob/de3ac9322ffc2fa0e35ccc94300e355ea3cdc6c5/pkg/storage/merge_queue.go#L258-L269
</blockquote></details>

Done
___
*[pkg/storage/client_merge_test.go, line 3424 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetT0Oe5g5O6jsy-PSK:-LewN3D38HKnbmlc_hxt:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/client_merge_test.go#L3424)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Idiomatic Go uses sentences (with proper punctuation) for all comments: https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences.
</blockquote></details>

Done
___
*[pkg/storage/client_merge_test.go, line 3430 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetTAEcE4fEuPrNYg_8:-Levq99c3MHqM38vBBKR:b-3bkcev) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/client_merge_test.go#L3430)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add a TODO to hook this up to the same mechanism that `ALTER TABLE ... UNSPLIT AT` uses when we eventually add that.
</blockquote></details>

I'm working on adding `ALTER TABLE ... UNSPLIT AT` in a followup commit. Should that be in a separate PR?
___
*[pkg/storage/merge_queue.go, line 213 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetTZptCxoq1iIiqI-t:-LewN-HX9UHXXTZH6SnO:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/merge_queue.go#L213)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Give this a comment.
</blockquote></details>

Done
___
*[pkg/storage/merge_queue.go, line 313 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetTKOU4nqTsf-VtAUj:-LewMz8d9bPYVSN76lvr:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/merge_queue.go#L313)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Please add a TODO to consider returning a "purgatory" error in this case to avoid repeatedly processing ranges that can't be merged.
</blockquote></details>

Done
___
*[pkg/storage/replica_command.go, line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetPFxX8rkJdoEf93du:-LewMvQWDVoWTCPNusay:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/storage/replica_command.go#L226)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is missing a `if args.Manual {` condition.

EDIT: it looks like you added this in the next commit. Let's rebase this fix back to the first one. Also, let's make sure that some test would have failed without this and without the corresponding condition in the main split transaction.
</blockquote></details>

Done
___
*[pkg/storage/replica_command.go, line 228 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetQQZQ5CBGcQxH5n1L:-LewMxGmEdQO2YX1ruOF:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/storage/replica_command.go#L228)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This transaction can currently race with a merge to leak sticky bits. In order to serialize with any merge transactions, I think we'll want to first read the range descriptor and make sure that it's equal to the provided `desc`. If not, we'll want to retry (see how `Replica.AdminSplit` handles `ConditionFailedError`s).

For reference, the split transaction below prevents these kinds of races by performing conditional puts on the range descriptor. We're not changing the range descriptor here, so there's no need to perform a CPut, but we'll want similar retry behavior.
</blockquote></details>

Done
___
*[pkg/storage/replica_command.go, line 549 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetTqwU866csTSv9sEe:-LewMqj66LfaArh0zATH:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/storage/replica_command.go#L549)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's get rid of this code movement. It changed in commit 1 and was later reverted. That could trip up a future `git blame`, so it's worth fixing in the first commit.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

jeffrey-xiao committed:
storage: add sticky bit key to range descriptor

The current design of automatic range merges must be aware of which
ranges were automatically or manually split since manual splits should
not be automatically merged. The sticky bit is a HLC added to the range
descriptor that indicates the range should be not merged into the merge
to its left.

The sticky bit is set on the RHS of a split operation. If you attempt to
split on the start key of a range, then the sticky bit of the range is
set, but no split occurs.

Release note: None

nvb reviewed (commented):
This looks good! Almost there. The only remaining question in my mind is whether we should represent the sticky bit as a boolean or as a timestamp. Let's discuss that inline below.

> Reviewed 86 of 86 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-:-Lf218AT77lfGqa3sGtT:b-etbo2p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jeffrey-xiao)

___
*[pkg/internal/client/db.go, line 469 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf1vICW9fj7TVoYZy7g:-Lf1vICX2EnFXLXWKa9Z:b-4qzxpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/internal/client/db.go#L469)):*
> ```Go
> // different range, hence the separation between spanKey and splitKey. See
> // #16008 for details, and #16344 for the tracking issue to clean this mess up
> // properly.
> ```

Add a paragraph about what the `manual` flag means.
___
*[pkg/roachpb/api.proto, line 639 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetRlH45vxo8CJ_-ufw:-Lf1y7IAA05C7slXuC7V:b-envd18) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/roachpb/api.proto#L639)):*
<details><summary><i>Previously, jeffrey-xiao (Jeffrey Xiao) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Right now we're just documenting how this is used, and not what the field actually does or why someone would want to set it.
___
*[pkg/roachpb/data.proto, line 198 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf1yXYK7cYoGiiW4uzc:-Lf1yXYK7cYoGiiW4uzd:b-39o4ou) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/roachpb/data.proto#L198)):*
> ```protobuf
> // StickyBitTrigger indicates that the sticky bit of a range has been changed.
> // This trigger is used in two cases:
> // 1. Unsplitting a range.
> ```

Define what we mean by "unsplitting" here. We don't want it to be confused with merging.
___
*[pkg/roachpb/metadata.proto, line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf1wBLv8yz1E-61cEBl:-Lf1wBLv8yz1E-61cEBm:b9hzslt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/roachpb/metadata.proto#L174)):*
> ```protobuf
>   // replica applying the merge is.
>   optional int64 generation = 6;
>   optional bool sticky_bit = 7;
> ```

We can add ` [(gogoproto.nullable) = false];` to this line to avoid the `*bool` in the generated code.
___
*[pkg/roachpb/metadata.proto, line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf1wM_xDe5lyUI_gHJJ:-Lf1wM_xDe5lyUI_gHJK:b4d63j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/roachpb/metadata.proto#L174)):*
> ```protobuf
>   // replica applying the merge is.
>   optional int64 generation = 6;
>   optional bool sticky_bit = 7;
> ```

Give this a comment. This will serve as the primary source of documentation on how the sticky bit works, what it's used for, and why it exists at all.
___
*[pkg/sql/split.go, line 109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf1z1kQ1-k4b5hb8ASv:-Lf1z1kQ1-k4b5hb8ASw:b-rwnr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/sql/split.go#L109)):*
> ```Go
> 	// because of gossip inconsistency, or it could change halfway through the
> 	// SPLIT AT's execution. It is, however, likely to prevent user error and
> 	// confusion in the common case.
> ```

Add a TODO to remove this in version 20.1.
___
*[pkg/sql/split.go, line 131 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf1zAD9CpgX7QAg7I2t:-Lf1zAD9CpgX7QAg7I2u:b6xhscl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/sql/split.go#L131)):*
> ```Go
> 	}
> 
> 	if err := params.extendedEvalCtx.ExecCfg.DB.AdminSplit(params.ctx, rowKey, rowKey, true /* manual */); err != nil {
> ```

Discussed in person. Let's set to the value of `params.EvalContext().Settings.Version.IsActive(cluster.VersionStickyBit)`.
___
*[pkg/storage/client_merge_test.go, line 3430 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetTAEcE4fEuPrNYg_8:-Lf1zqME4bKdOo3o_lsw:b-62bo3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/client_merge_test.go#L3430)):*
<details><summary><i>Previously, jeffrey-xiao (Jeffrey Xiao) wrote…</i></summary><blockquote>

I'm working on adding `ALTER TABLE ... UNSPLIT AT` in a followup commit. Should that be in a separate PR?
</blockquote></details>

Yes, this should be in a follow-up PR.
___
*[pkg/storage/replica_command.go, line 319 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetRMP1DoX2MprXvgFd:-Lf2-XH697gLQ8nPslmP:bl1kcax) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/storage/replica_command.go#L319)):*
<details><summary><i>Previously, jeffrey-xiao (Jeffrey Xiao) wrote…</i></summary><blockquote>

I think it depends if we want to support temporarily not merging manual splits. We would also need to consider how this time frame can be expressed on the SQL side.
</blockquote></details>

Intuitively I would lean towards making this a timestamp so that we could add an expiration policy to manual splits in the future for without a tricky migration, but doing so will come with a cost. A bool will encode to a single byte while an HLC timestamp will encode to ~8.

@ajwerner I'm curious if you have an opinion on this.
___
*[pkg/storage/replica_command.go, line 228 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf21Zut8mD6xGjEg7yX:-Lf21Zut8mD6xGjEg7yY:bhzzr5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/storage/replica_command.go#L228)):*
> ```Go
> 		log.Event(ctx, "range already split")
> 		// Even if the range is already split, we should still set the sticky bit
> 		if args.Manual {
> ```

`if args.Manual && !desc.StickyBit {`, right?
___
*[pkg/storage/replica_command.go, line 233 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf21UF9BgEYIIANjNGG:-Lf21UF9BgEYIIANjNGH:bakyupk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/storage/replica_command.go#L233)):*
> ```Go
> 			err := r.store.DB().Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
> 				b := txn.NewBatch()
> 				if err := updateRangeDescriptor(b, keys.RangeDescriptorKey(desc.StartKey), desc, &newDesc); err != nil {
> ```

nit: pull `descKey := keys.RangeDescriptorKey(desc.StartKey)` out into the previous line.


<!-- Sent from Reviewable.io -->

jeffrey-xiao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37506)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @nvanbenschoten)

___
*[pkg/roachpb/metadata.proto, line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf1wBLv8yz1E-61cEBl:-Lf5Kkvv-8p-9d66cUgz:b-o2k83q) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/roachpb/metadata.proto#L174)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We can add ` [(gogoproto.nullable) = false];` to this line to avoid the `*bool` in the generated code.
</blockquote></details>

I don't think I can set `[(gogoproto.nullable) = false];` for native fields in proto3. I'm getting an error when attempting to set it, but please correct me if I'm wrong.
___
*[pkg/storage/client_merge_test.go, line 3430 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetTAEcE4fEuPrNYg_8:-Lf5L2CEEF3ekaeoX2pw:brxp9sd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d5b8ed64e0d1a215c03594f492104385a522a8/pkg/storage/client_merge_test.go#L3430)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yes, this should be in a follow-up PR.
</blockquote></details>

Sounds good!


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37506)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jeffrey-xiao, and @nvanbenschoten)

___
*[pkg/roachpb/metadata.proto, line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf1wBLv8yz1E-61cEBl:-Lf6BfJxDtpkWIirLnLf:bxcbj84) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/roachpb/metadata.proto#L174)):*
<details><summary><i>Previously, jeffrey-xiao (Jeffrey Xiao) wrote…</i></summary><blockquote>

I don't think I can set `[(gogoproto.nullable) = false];` for native fields in proto3. I'm getting an error when attempting to set it, but please correct me if I'm wrong.
</blockquote></details>

What error are you getting? This file isn't proto3, it's proto2.


<!-- Sent from Reviewable.io -->

jeffrey-xiao commented:
___
*[pkg/roachpb/metadata.proto, line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf1wBLv8yz1E-61cEBl:-Lf6Cq7D3UM8_xhno6Qt:bmhlgi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/roachpb/metadata.proto#L174)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What error are you getting? This file isn't proto3, it's proto2.
</blockquote></details>

Oh this file is proto2, but I also wanted to keep it consistent with StickyBitTrigger which is a proto3 file. The error I'm getting is:

```
ERROR: field StickyBitTrigger.StickyBit is a native type and in proto3 syntax with nullable=false there exists conflicting implementations when encoding zero values--gogoroach_out: protoc-gen-gogoroach: Plugin failed with status code 1.
```

Another place I'm using a boolean is the Manual flag in AdminSplitRequest.

Should I just add the nullable attribute to only the sticky_bit in RangeDescriptor?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37506)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jeffrey-xiao and @nvanbenschoten)

___
*[pkg/storage/replica_command.go, line 319 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LetRMP1DoX2MprXvgFd:-Lf6CCgu8XGAI1HBAgF7:b-wy59ov) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1488d446ca9e9c1fe2facf1368e1e3f0628c7d/pkg/storage/replica_command.go#L319)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Intuitively I would lean towards making this a timestamp so that we could add an expiration policy to manual splits in the future for without a tricky migration, but doing so will come with a cost. A bool will encode to a single byte while an HLC timestamp will encode to ~8.

@ajwerner I'm curious if you have an opinion on this.
</blockquote></details>

I'm on board with the HLC. Given that start and end keys live in range descriptors and they can be quite large it's probably fine. It will probably also help from a UX perspective. I feel like if I come to a cluster and find it has manual splits I'd be curious about when they were added. 


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 18 of 18 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-:-Lf6v9nW8sLRl1-R7d0I:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jeffrey-xiao)

___
*[pkg/roachpb/data.proto, line 207 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf6uuwSCsfAEg5TPiqd:-Lf6uuwSCsfAEg5TPiqe:b-o7m56e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c35801e47e4086bc5e83a8cb48c66296b9600b7/pkg/roachpb/data.proto#L207)):*
> ```protobuf
>   option (gogoproto.equal) = true;
> 
>   util.hlc.Timestamp sticky_bit = 1;
> ```

Add a comment along the lines of `// Set to nil to remove a RangeDescriptor's sticky bit.`


<!-- Sent from Reviewable.io -->

jeffrey-xiao commented:
___
*[pkg/storage/replica_command.go, line 228 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf21Zut8mD6xGjEg7yX:-LfKXiBVEpU5KgpqPUom:b-oi8gwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/storage/replica_command.go#L228)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`if args.Manual && !desc.StickyBit {`, right?
</blockquote></details>

Since the sticky bit is changed to a HLC, we should check for `desc.StickyBit == nil` anymore because we would want to update the HLC right?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37506)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jeffrey-xiao)

___
*[pkg/storage/replica_command.go, line 228 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf21Zut8mD6xGjEg7yX:-LfKYcWa7if3BzG_7Ir4:b-otaepx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b720dd34de23314eb542bf0e7b763fc3c151571a/pkg/storage/replica_command.go#L228)):*
<details><summary><i>Previously, jeffrey-xiao (Jeffrey Xiao) wrote…</i></summary><blockquote>

Since the sticky bit is changed to a HLC, we should check for `desc.StickyBit == nil` anymore because we would want to update the HLC right?
</blockquote></details>

That's an interesting point. So you're saying that we should "refresh" the sticky bit to a higher timestamp? Yes, sounds like the right behavior to me. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> really nice work! Update the comments on the SickyBitTrigger and merge it!

> Reviewed 1 of 5 files at r2, 66 of 86 files at r3, 15 of 18 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-:-LfK_wcnBGWDnkcch-et:b-rjvlde)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @jeffrey-xiao, and @nvanbenschoten)

___
*[pkg/roachpb/data.proto, line 196 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-LfKYpYB4kcPGp5uwHWk:-LfKYpYC9m9I2YLIldc6:bjjwcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c35801e47e4086bc5e83a8cb48c66296b9600b7/pkg/roachpb/data.proto#L196)):*
> ```protobuf
> }
> 
> // StickyBitTrigger indicates that the sticky bit of a range has been changed.
> ```

nit: s/has been/should be/?

Is there any validation that this actually is the timestamp that's in the range descriptor when this commits? It's probably worth a comment that this is always set to the value used in a CPut for the transaction and that it's the client's responsibility to ensure that the timestamps align. 
___
*[pkg/roachpb/data.proto, line 207 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37506#-Lf6uuwSCsfAEg5TPiqd:-LfKZ_rE0NQ-LGwQv_YD:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c35801e47e4086bc5e83a8cb48c66296b9600b7/pkg/roachpb/data.proto#L207)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add a comment along the lines of `// Set to nil to remove a RangeDescriptor's sticky bit.`
</blockquote></details>

+1


<!-- Sent from Reviewable.io -->

jeffrey-xiao commented:
bors r+
